### PR TITLE
Adding blobstore package which supports Swift

### PR DESCRIPTION
**Pull request by dhlambert**
Extension to GWC which adds native support for Swift. This allows OpenStack Cloud deployments to easily use GeoWebCache without the support from external libraries such as S3Proxy. This has been implemented using the open source multi-cloud toolkit jclouds (https://jclouds.apache.org/).

**Additions:**
- Swift Blobstore package containing blobstore class and supporting config provider classes
- Additions to configuration and pom files for swift
- Unit tests covering blobstore class

### CODE REVIEW

**Comment by dhlambert** [comment]
Seems the the output of the failing Travis tests produce log output at the debug level when building the web module. Have had a look into this but still not sure why this is occurring and what changes might have caused this?

File: geowebcache/swiftblob/src/main/java/org/geowebcache/swift/SwiftBlobStoreConfigProvider.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+ * of the License, or (at your option) any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * <p>You should have received a copy of the GNU Lesser General Public License along with this
+ * program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * @author Gabriel Roldan, Boundless Spatial Inc, Copyright 2015
```
Comment (jodygarnett) [inline geowebcache/swiftblob/src/main/java/org/geowebcache/swift/SwiftBlobStoreConfigProvider.java]: This is a new file, so I expect you are the author copyright 2020?

File: geowebcache/swiftblob/src/main/java/org/geowebcache/swift/SwiftBlobStoreConfigProvider.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+import com.thoughtworks.xstream.converters.basic.BooleanConverter;
+import com.thoughtworks.xstream.converters.basic.IntConverter;
+import com.thoughtworks.xstream.converters.basic.StringConverter;
+import org.geowebcache.GeoWebCacheEnvironment;
+import org.geowebcache.GeoWebCacheExtensions;
+import org.geowebcache.config.BlobStoreInfo;
+import org.geowebcache.config.Info;
+import org.geowebcache.config.XMLConfigurationProvider;
+
+public class SwiftBlobStoreConfigProvider implements XMLConfigurationProvider {
```
Comment (jodygarnett) [inline geowebcache/swiftblob/src/main/java/org/geowebcache/swift/SwiftBlobStoreConfigProvider.java]: You may want a public javadoc

File: geowebcache/swiftblob/src/main/java/org/geowebcache/swift/TMSKeyBuilder.java [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+ * of the License, or (at your option) any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * <p>You should have received a copy of the GNU Lesser General Public License along with this
+ * program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * @author Gabriel Roldan, Boundless Spatial Inc, Copyright 2015
```
Comment (jodygarnett) [inline geowebcache/swiftblob/src/main/java/org/geowebcache/swift/TMSKeyBuilder.java]: as above

**Comment by dhlambert** [comment]
@jodygarnett I have updated javadoc and added some documentation in  the configuration/storage.rst file for Openstack Swift Blobstore.  Have also made some refactoring changes within config for some unused properties.

I have not updated the author as the TMSKeyBuilder class is unmodified apart from javadoc, copied from s3storage - see #830. Also the SwiftBlobStoreConfigProvider.java file has some untouched code and some added code. Not sure if I should be taking all the credit for another persons work.

Commits have been left unsquashed for reviewing.

**Comment by dhlambert** [comment]
Hi @jodygarnett,

Just wondering if you think this pull request needs any further changes? Also, if there is any proposed timeline for merging?

**Comment by jodygarnett** [comment]
> Hi @jodygarnett,
> 
> Just wondering if you think this pull request needs any further changes? Also, if there is any proposed timeline for merging?

Let me check, and yes I am happy to merge.

Please note a release candidate went out last week, after the release goes out we can backport this work also.

File: documentation/en/user/source/configuration/storage.rst [hunk 4]
```diff
@@ -791,3 +791,63 @@ In this section are described other available configuration parameters to config
 		
 		.. note:: A value of *max-size* bigger or equal to Integer.MAX_VALUE cannot be used in order to avoid an uncontrollable growth of the cache size.
 
+OpenStack Swift (Swift) Blob Store
++++++++++++++++++++++++++++++++++++++++++++++
```
Comment (jodygarnett) [inline documentation/en/user/source/configuration/storage.rst]: Thanks this is perfect

File: geowebcache/swiftblob/pom.xml [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        <dependency>
+            <groupId>org.geowebcache</groupId>
+            <artifactId>gwc-core</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <!-- jclouds OpenStack dependencies -->
+        <dependency>
+            <groupId>org.apache.jclouds.api</groupId>
+            <artifactId>openstack-swift</artifactId>
+            <version>${jclouds.version}</version>
```
Comment (jodygarnett) [inline geowebcache/swiftblob/pom.xml]: We try and add version numbers to the root pom.xml dependency management section; something to consider for later.

File: geowebcache/swiftblob/pom.xml [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-test</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+            <version>3.2.4</version>
```
Comment (jodygarnett) [inline geowebcache/swiftblob/pom.xml]: Please use dependency management section to manage version numbers (so we do not end up with conflicts). 
This is a test dependency so you will not get a conflict in this case.

File: geowebcache/swiftblob/pom.xml [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
+
+    <build>
+        <plugins>
```
Comment (jodygarnett) [inline geowebcache/swiftblob/pom.xml]: Q: Why empty plugins section?

**Pull request merged**